src/tir/transforms/storage_flatten.cc

```diff
@@ -50,6 +50,913 @@ using runtime::StorageRank;
+      // Keep the extent of the buffer view the same.
+      new_tuple_args.push_back(view_extent);
+      // Use the extent of the buffer view to define the buffer view's shape.
+      realized_shape.push_back(view_extent);
+      // Within the buffer view, indices start at 0.
+      realized_begins.push_back(0);
+    }
+
+    Buffer key = buffer;
+
+    auto write_ptr = buffer.CopyOnWrite();
+    write_ptr->shape = realized_shape;
+
+    {
+      InternalBufferRemap remap;
+      remap.realized_begins = realized_begins;
+      remap.remap_to = buffer;
+      remap.in_scope = true;
+      internal_buf_map_[key] = remap;
+    }
+
+    Stmt stmt = AttrStmt(Array<ObjectRef>{buffer, target_remap.remap_to}, op->attr_key,
+                         Call(tuple->dtype, tuple->op, new_tuple_args, tuple->span),
+                         this->VisitStmt(op->body));
+    internal_buf_map_.at(key).in_scope = false;
+    return stmt;
+  }
+
+  std::unordered_set<Buffer, ObjectPtrHash, ObjectPtrEqual> extern_buffers_;
+
+  struct InternalBufferRemap {
+    Buffer remap_to;
+    Array<PrimExpr> realized_begins;
+    bool in_scope;
+  };
+
+  std::unordered_map<Buffer, InternalBufferRemap, ObjectPtrHash, ObjectPtrEqual> internal_buf_map_;
+
+  IRVisitorWithAnalyzer* bound_analyzer_;
+};
+
+/* Apply dimension alignment restrictions
+ *
+ * Buffers annotated with attr::buffer_dim_align may need to have
+ * strides defined such that they are no longer in a compact shape.
+ * After this pass, buffers have stride definitions to include these
+ * alignment restrictions, and attr::buffer_dim_align annotations have
+ * been removed.
+ */
+class BufferStrideLegalize : public StmtExprMutator {
+ public:
+  explicit BufferStrideLegalize(const Map<Var, Buffer>& extern_buffer_map,
+                                IRVisitorWithAnalyzer* bound_analyzer)
+      : bound_analyzer_(bound_analyzer) {
+    for (auto kv : extern_buffer_map) {
+      Buffer buf = kv.second;
+      Buffer with_strides = WithStrides(buf);
+      {
+        BufferEntry entry;
+        entry.remap_to = with_strides;
+        entry.in_scope = true;
+        entry.is_external = true;
+        buf_map_[buf] = entry;
+      }
+      updated_extern_buffer_map_.Set(kv.first, with_strides);
+    }
+  }
+
+  Map<Var, Buffer> UpdatedExternBufferMap() const { return updated_extern_buffer_map_; }
+
+  Buffer WithStrides(Buffer buf) {
+    auto it = buf_map_.find(buf);
+    if (it != buf_map_.end()) {
+      const BufferEntry& entry = it->second;
+      ICHECK(entry.in_scope) << "Cannot annotate an out-of-scope buffer";
+      return entry.remap_to;
+    }
+
+    if (buf->strides.size()) {
+      ICHECK_EQ(buf->strides.size(), buf->shape.size());
+      return buf;
+    }
+
+    Array<PrimExpr> shape = buf->shape;
+
+    // Keeping this to have matched behavior to previous version.
+    // There are many parts of the codebase that assume that a strided
+    // array cannot be compact.
```

Comment:
nit: An example of one such case that makes this assumption could be useful for context.

Reply:
Makes sense, and added.